File: solutions/rust/problem1_max_non_adjacent_sum.rs
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();
    let n: usize = match it.next() { Some(x) => x.parse().unwrap(), None => return };
    let mut w = vec![0i64; n+1];
    for i in 1..=n { w[i] = it.next().unwrap().parse().unwrap(); }
    let mut g = vec![Vec::new(); n+1];
    for _ in 0..n-1 {
        let u: usize = it.next().unwrap().parse().unwrap();
        let v: usize = it.next().unwrap().parse().unwrap();
        g[u].push(v); g[v].push(u);
    }

    let mut parent = vec![0usize; n+1];
    let mut order = vec![1usize];
    for i in 0..order.len() {
        let v = order[i];
        for &to in &g[v] { if to != parent[v] { parent[to] = v; order.push(to); } }
    }

    let mut dp0 = vec![0i64; n+1];
    let mut dp1 = vec![0i64; n+1];

    for &v in order.iter().rev() {
        let mut select = w[v];
        let mut notsel = 0i64;
        for &to in &g[v] { if to != parent[v] { select += dp0[to]; notsel += dp0[to].max(dp1[to]); } }
        dp0[v] = notsel; dp1[v] = select;
    }

    println!("{}", dp0[1].max(dp1[1]));
}

File: solutions/go/problem1_max_non_adjacent_sum.go
package main

import (
    "bufio"
    "fmt"
    "os"
)

func main() {
    in := bufio.NewReader(os.Stdin)
    var n int
    if _, err := fmt.Fscan(in, &n); err != nil {
        return
    }
    w := make([]int64, n+1)
    for i := 1; i <= n; i++ {
        fmt.Fscan(in, &w[i])
    }
    g := make([][]int, n+1)
    for i := 0; i < n-1; i++ {
        var u, v int
        fmt.Fscan(in, &u, &v)
        g[u] = append(g[u], v)
        g[v] = append(g[v], u)
    }

    dp0 := make([]int64, n+1)
    dp1 := make([]int64, n+1)

    parent := make([]int, n+1)
    order := []int{1}
    for i := 0; i < len(order); i++ {
        v := order[i]
        for _, to := range g[v] {
            if to != parent[v] {
                parent[to] = v
                order = append(order, to)
            }
        }
    }

    for i := n - 1; i >= 0; i-- {
        v := order[i]
        selectSum := w[v]
        notSel := int64(0)
        for _, to := range g[v] {
            if to != parent[v] {
                selectSum += dp0[to]
                if dp0[to] > dp1[to] {
                    notSel += dp0[to]
                } else {
                    notSel += dp1[to]
                }
            }
        }
        dp0[v] = notSel
        dp1[v] = selectSum
    }

    if dp0[1] > dp1[1] {
        fmt.Println(dp0[1])
    } else {
        fmt.Println(dp1[1])
    }
}

